### PR TITLE
Statman and ICMP

### PR DESCRIPTION
**Pull request by AnnikaH**

### CODE REVIEW

**Comment by AnnikaH** [comment]
Jenkins test please

**Comment by AnnikaH** [comment]
Jenkins test please

File: api/net/ip4/icmp4.hpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    /**
+     *  Find the ping-callback that this packet is a response to, execute it and erase the object
+     *  from the ping_callbacks_ map
+     */
+    inline void execute_ping_callback(icmp4::Packet& ping_response) {
+      // Find callback matching the reply
+      auto it = ping_callbacks_.find(std::make_pair(ping_response.id(), ping_response.sequence()));
+
+      if (it != ping_callbacks_.end()) {
+        it->second.callback(ICMP_packet{ping_response.id(), ping_response.sequence(), ping_response.ip().ip_src(),
```
Comment (AndreasAakesson) [inline api/net/ip4/icmp4.hpp]: To make it cleaner you could add a constructor to `struct ICMP_packet` that takes a `const icmp4::Packet&` and hide all these lines there (with a shorter name).

File: api/net/ip4/icmp4_common.hpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    enum class Parameter_problem : uint8_t {
+      POINTER_INDICATES_ERROR,
+      REQUIRED_OPT_MISSING      // RFC 1122
+    };
+
+  } // < namespace code
+
+  static std::string __attribute__((unused)) get_type_string(Type type) {
+    return [&type] () {
```
Comment (AndreasAakesson) [inline api/net/ip4/icmp4_common.hpp]: This lambda doesn't fill any purpose here.

File: api/net/ip4/icmp4_common.hpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        case Type::TIMESTAMP_REPLY:
+          return "TIMESTAMP REPLY (14)";
+        case Type::NO_REPLY:
+          return "NO REPLY";
+      };
+    }();
+  }
+
+  static std::string __attribute__((unused)) get_code_string(Type type, uint8_t code) {
+    return [&type, &code] () {
```
Comment (AndreasAakesson) [inline api/net/ip4/icmp4_common.hpp]: Same as above. 
Maybe have helper functions for the nested switch cases? It also could make sense to move this to a cpp file.

File: src/net/ip4/icmp4.cpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    }
+  }
+
+  void ICMPv4::forward_to_transport_layer(icmp4::Packet& req) {
+    // inet forwards to transport layer (UDP or TCP)
+    inet_.error_report(req.type(), req.code(), req.payload());
+  }
+
+  void ICMPv4::destination_unreachable(Packet_ptr pckt, icmp4::code::Dest_unreachable code) {
+    if ((size_t)pckt->size() < sizeof(IP4::header) + icmp4::Packet::header_size()) // Drop if not a full header
```
Comment (AndreasAakesson) [inline src/net/ip4/icmp4.cpp]: You could make it more DRY by creating a inline helper function for this, and also use UNLIKELY:
`if(UNLIKELY(not full_header(*pckt))) return;`
(Could this be done before any of the functions are called?)

File: src/net/ip4/udp.cpp [hunk 5]
```diff
@@ -34,22 +35,49 @@ namespace net {
...
+    // But only if the destination IP address is not broadcast or multicast
+    auto ip4_packet = static_unique_ptr_cast<PacketIP4>(std::move(udp_packet));
+    if (ip4_packet->ip_dst() != IP4::ADDR_BCAST and (ip4_packet->ip_dst().part(3) <= 224 or
+      ip4_packet->ip_dst().part(3) >= 239))
+    {
+      stack_.icmp().destination_unreachable(std::move(ip4_packet), icmp4::code::Dest_unreachable::PORT);
+    }
+  }
+
+  void UDP::error_report(Error_type type, Error_code code,
```
Comment (alfred-bratterud) [inline src/net/ip4/udp.cpp]: Signature will change to e.g. `UDP::error_report(ICMP_Error)` or something, as discussed

File: src/net/tcp/tcp.cpp [hunk 6]
```diff
@@ -196,6 +196,13 @@ string TCP::to_string() const {
   return ss.str();
 }
 
+void TCP::error_report(Error_type type, Error_code code,
```
Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: New signature as discussed for UDP

File: test/net/integration/icmp/service.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  // Waiting 30 seconds for reply
+  inet.icmp().ping(IP4::addr{10,0,0,43}, [](ICMP_packet pckt) {
+    if (pckt)
+      printf("Received packet from 10.0.0.43\n%s\n", pckt.to_string().c_str());
+    else
+      printf("No reply received from 10.0.0.43\n");
+  });
+
+  // Also possible without callback:
+  // inet.icmp().ping(IP4::addr{193,90,147,109});  // google.com
```
Comment (alfred-bratterud) [inline test/net/integration/icmp/service.cpp]: Would also be nice to do DNS resolution - so you could `ping("google.com", callback)`. For later.

**Pull request merged**